### ISSUE TITLE
Nullable object must have a value. SingleQueryResultCoordinator.

### ISSUE THREAD

**Issue by skclusive** [dotnet/efcore#22517]
when Blog does not have a Post, following query does not work in 5.0.0-preview.8.* or 6.0.0-* nightly. but works in 5.0.0-preview.7.*

```c#

 public class Blog
    {
        public int BlogId { get; set; }
        public string Url { get; set; }

        [NotMapped]
        public Post Post { get; set; }

        public List<Post> Posts { get; set; } = new List<Post>();
    }

    public class Post
    {
        public int PostId { get; set; }

        public string Title { get; set; }

        public string Content { get; set; }

        public int BlogId { get; set; }

        public Blog Blog { get; set; }
    }

// this IQueryable would come from other API.

 var dbPosts = from p in db.Posts
                          // select p;
                         select new Post
                         {
                             PostId = p.PostId,
                             BlogId = p.BlogId,
                             Content = p.Content
                         };

 var query = from blog in db.Blogs
                    join post in dbPosts on blog.BlogId equals post.BlogId into posts
                    from xpost in posts.DefaultIfEmpty()
                    select new Blog
                    {
                         Url = blog.Url,
                         Post = xpost
                   };

```

### Steps to reproduce

I have a repo to reproduce the bug.

https://github.com/skclusive/EFLeftJoinBug

Unhandled exception. System.InvalidOperationException: Nullable object must have a value.
   at lambda_method17(Closure , QueryContext , DbDataReader , ResultContext , SingleQueryResultCoordinator )
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.Enumerator.MoveNext()

### Further technical details

EF Core version:
Database provider: (e.g. Microsoft.EntityFrameworkCore.Sqlite)
Target framework: (e.g. .NET Core 5.0)
Operating system:
IDE: (e.g. Visual Studio Code)

**Comment by ajcvickers** [comment]
Confirmed this works in 3.1, but fails in latest daily, on SQL Server and SQLite.

**Comment by smitpatel** [comment]
@ajcvickers - What is the generated SQL? Can you post query plan?

**Comment by ajcvickers** [comment]
@smitpatel Here's the logs:

```
dbug: Microsoft.EntityFrameworkCore.Query[10111]
      Compiling query expression: 
      'DbSet<Blog>()
          .GroupJoin(
              inner: DbSet<Post>()
                  .Select(p => new Post{ 
                      PostId = p.PostId, 
                      BlogId = p.BlogId, 
                      Content = p.Content 
                  }
                  ), 
              outerKeySelector: blog => blog.BlogId, 
              innerKeySelector: post => post.BlogId, 
              resultSelector: (blog, posts) => new { 
                  blog = blog, 
                  posts = posts
               })
          .SelectMany(
              collectionSelector: <>h__TransparentIdentifier0 => <>h__TransparentIdentifier0.posts
                  .DefaultIfEmpty(), 
              resultSelector: (<>h__TransparentIdentifier0, xpost) => new Blog{ 
                  Url = <>h__TransparentIdentifier0.blog.Url, 
                  Post = xpost 
              }
          )'
dbug: Microsoft.EntityFrameworkCore.Query[10107]
      queryContext => new SingleQueryingEnumerable<Blog>(
          (RelationalQueryContext)queryContext, 
          RelationalCommandCache.SelectExpression(
              Projection Mapping:
                  Url -> 0
                  Post.PostId -> 1
                  Post.BlogId -> 2
                  Post.Content -> 3
              SELECT b.Url, p.PostId, p.BlogId, p.Content
              FROM Blogs AS b
              LEFT JOIN Posts AS p ON b.BlogId == p.BlogId), 
          Func<QueryContext, DbDataReader, ResultContext, SingleQueryResultCoordinator, Blog>, 
          EFLeftJoinBug.BloggingContext, 
          False, 
          False
      )
dbug: Microsoft.EntityFrameworkCore.Database.Command[20103]
      Creating DbCommand for 'ExecuteReader'.
dbug: Microsoft.EntityFrameworkCore.Database.Command[20104]
      Created DbCommand for 'ExecuteReader' (0ms).
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20000]
      Opening connection to database 'Test' on server '(local)'.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20001]
      Opened connection to database 'Test' on server '(local)'.
dbug: Microsoft.EntityFrameworkCore.Database.Command[20100]
      Executing DbCommand [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT [b].[Url], [p].[PostId], [p].[BlogId], [p].[Content]
      FROM [Blogs] AS [b]
      LEFT JOIN [Posts] AS [p] ON [b].[BlogId] = [p].[BlogId]
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (7ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT [b].[Url], [p].[PostId], [p].[BlogId], [p].[Content]
      FROM [Blogs] AS [b]
      LEFT JOIN [Posts] AS [p] ON [b].[BlogId] = [p].[BlogId]
fail: Microsoft.EntityFrameworkCore.Query[10100]
      An exception occurred while iterating over the results of a query for context type 'EFLeftJoinBug.BloggingContext'.
      System.InvalidOperationException: Nullable object must have a value.
         at lambda_method(Closure , QueryContext , DbDataReader , ResultContext , SingleQueryResultCoordinator )
         at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.Enumerator.MoveNext()
System.InvalidOperationException: Nullable object must have a value.
   at lambda_method(Closure , QueryContext , DbDataReader , ResultContext , SingleQueryResultCoordinator )
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.Enumerator.MoveNext()
Unhandled exception. System.InvalidOperationException: Nullable object must have a value.
   at lambda_method(Closure , QueryContext , DbDataReader , ResultContext , SingleQueryResultCoordinator )
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.Enumerator.MoveNext()
   at EFLeftJoinBug.Program.Main(String[] args) in /home/ajcvickers/AllTogetherNow/Daily/Daily.cs:line 43
```

**Comment by smitpatel** [comment]
SingleQueryResultCoordinator is not the error. We are probably trying to materialize null into something which is non-nullable which should have been skipped from materialization.

**Comment by smitpatel** [comment]
Note: This gave incorrect result in 3.1.x by initialize post with default values rather than returning null (the result of DefaultIfEmpty).

With #20633 fixed, it now throws error.

Right fix would be not to create instance if no rows matched.

**Comment by smitpatel** [comment]
Issues faced:
When generating left join we need to add additional null check in selector so that we can generate correct default type if sequence is empty and not try to materialize result.
- This causes a complex selector which when used inside anything other than select, fails to translate since we don't bind to column eventually. (it is ternary with client side types). On flip side non-select shouldn't need check either since column would be null anyway.
- We cannot do when translating LeftJoin easily because projection is lifted after LeftJoin so we don't have enough data in translation pipeline in that sense.
- For unexpanded navigations we try to add check which is not necessary. Requires #20291

Due to unintended side effect of breaking working queries due to having a complex expression for entity now, we should postpone this.

**Comment by smitpatel** [comment]
Work-around to original issue
```
 var query = from blog in db.Blogs
                    join post in dbPosts on blog.BlogId equals post.BlogId into posts
                    from xpost in posts.DefaultIfEmpty()
                    select new Blog
                    {
                         Url = blog.Url,
                         Post = xpost == null ? null : xpost
                   };
```

**Comment by skclusive** [comment]
i did try the following work-around before raising the bug. this workaround doesn't work :(

```c#
 var query = from blog in db.Blogs
                    join post in dbPosts on blog.BlogId equals post.BlogId into posts
                    from xpost in posts.DefaultIfEmpty()
                    select new Blog
                    {
                         Url = blog.Url,
                         Post = xpost == null ? null : xpost
                   };
```

**Comment by ajcvickers** [comment]
@skclusive Did you try the workaround using EF Core 5.0 RC1 or EF Core 3.1.x?

**Comment by skclusive** [comment]
@ajcvickers i tested with 6.0.0-* nightly. now tried with EF Core RC1 also. same issue.

**Comment by skclusive** [comment]
i guess this is critical issue.

if i use in memory objects. following does work with standard linq query. so i am wondering why #20633 considered bug.

```c#
var memoryBlogs = new List<Blog>
                {
                    new Blog
                    {
                        BlogId = 1,
                        Url = "http://blogs.msdn.com/adonet"
                     }
                };

                var memoryPosts = from p in new List<Post>()
                            select new Post
                            {
                                PostId = p.PostId,
                                BlogId = p.BlogId,
                                Content = p.Content
                            };

                var query = from blog in memoryBlogs
                            join post in memoryPosts on blog.BlogId equals post.BlogId into posts
                            from xpost in posts.DefaultIfEmpty()
                            select new Blog
                            {
                                Url = blog.Url,
                                Post = xpost
                            };

```

when there is no posts, following does not throw error in the projection. so why throwing error when composed.


```c#

var posts = from p in db.Posts
                            select new Post
                            {
                                PostId = p.PostId, // we dont consider p null here
                                BlogId = p.BlogId,
                                Content = p.Content
                            };

                var post = posts.SingleOrDefault(); // query is success with null value. no error on projection.

                Console.WriteLine(post?.PostId);
```

i am composing IQueryable<T> extensively in my code and this recent change breaks my framework. so please consider to address this issue.

**Comment by smitpatel** [comment]
@skclusive - Your query was giving incorrect results in 3.1, The bug fix for #20633 stopped generating incorrect results. In your case it throws exception because the query translation couldn't figure out a way to generate right results. Your query working in LINQ is irrelevant as it worked in 3.1 also and EF Core generated results different than LINQ.

```C#
                var dbPosts = from p in db.Posts
                                  // select p;
                              select new Post
                              {
                                  PostId = p.PostId,
                                  BlogId = p.BlogId,
                                  Content = p.Content
                              };

                var query = from blog in db.Blogs
                            join post in dbPosts on blog.BlogId equals post.BlogId into posts
                            from xpost in posts.DefaultIfEmpty()
                            select new Blog
                            {
                                Url = blog.Url,
                                Post = xpost.BlogId == null ? null : xpost
                            };
```

Tested that above work-around gives correct result on 5.0 rc2 nightly build.

**Comment by skclusive** [comment]
@smitpatel got it. thanks. 

the above mentioned workaround does work in nightly and rc-1.

only issue is this comparison produce warning xpost.BlogId == null (comparing int to null). Also will this be documented as i guess some might get this issue frequently.

being workaround will this be addressed in future releases?

you can close the issue if no further action involved.

**Comment by shadowfoxish** [comment]
https://stackoverflow.com/a/65207398/1181624 

You can also do a cast to a nullable type to make this issue go away. I posted that stack overflow answer before I found this post.

**Comment by hhhuut** [comment]
Still happens with MetadataGenerator v1.2.0 and the latest EFC packages (5.0.1).

However from what I understand, this is now the expected behaviour and we'll have to add explicit "workarounds" to all left-joined queries like the one @shadowfoxish mentioned?

**Comment by smitpatel** [comment]
@hhhuut - There is no one answer to the question.
Essentially, when you do a left join, right side can have null values so accessing any property over that will give you back null (EF Core will make property access null safe). But in C# there is no concept of left join or null-safe access. So in C# the property access will take CLR type of the property which can/cannot accommodate null value.
e.g. `entity.Property` where entity is coming from right side of left join and `Property` is of type int will return int value. If entity turns out to be null then it throws null-ref exception. Since EF Core makes the access null safe it is same as writing `entity == null ? (int?)null : entity.Property` which will make the return value of type `int?` So when actual null value is encountered then it cannot be put into a int value throwing above exception. This only happens at runtime, when you actually encounter a null value.

So, as a user, you should carefully evaluate queries and see if you are expecting a null value appearing in left join and the property-accesses are able to accommodate that null value by being nullable type. In such cases you are required to put nullable type cast. It is not a "work-around", it is required to materialize the results from the data you have. Though if your query will never get null from left join (be careful as data can change over the type in an app), or you are not accessing a property on an entity which can be nullable due to left join then you shouldn't need it. The same exception can also arise when database contains null but client side property cannot take null values (even without left-joins) due to mismatch in model. There is also a possibility that there is a bug in EF Core somewhere. So if you have a query which has result types correct based on expected data and seeing this exception please file a new issue.

**Comment by cgountanis** [comment]
Just seems like a pain if you have a table with optional FK. When you use .Select() to optimize it just blows up with "Nullable object must have a value."

```
return _context.Something
	.OrderByDescending(j => j.X)
	.Include(r => r.X)
	.Include(r => r.XX)
	
	.Select(r => new Something()
	{
		X1 = r.X1,
		X2 = r.X2,

		X = new X() 
		{
			Name= r.X.Name
		},
		XX = r.XXNullableId == null ? null : new XX()
		{
			XNumber = r.XX.XNumber
		}
	})
	.AsSingleQuery()
	.AsNoTracking();
```

You end up getting this strange CASE statement but it seems to work fine this way.
```
CASE WHEN [d].[XXId] IS NULL THEN CAST(1 AS bit) ELSE CAST(0 AS bit)
```


Is this the best approach?

**Comment by jayeshdshah** [comment]
> @hhhuut - There is no one answer to the question.
> Essentially, when you do a left join, right side can have null values so accessing any property over that will give you back null (EF Core will make property access null safe). But in C# there is no concept of left join or null-safe access. So in C# the property access will take CLR type of the property which can/cannot accommodate null value.
> e.g. `entity.Property` where entity is coming from right side of left join and `Property` is of type int will return int value. If entity turns out to be null then it throws null-ref exception. Since EF Core makes the access null safe it is same as writing `entity == null ? (int?)null : entity.Property` which will make the return value of type `int?` So when actual null value is encountered then it cannot be put into a int value throwing above exception. This only happens at runtime, when you actually encounter a null value.
> 
> So, as a user, you should carefully evaluate queries and see if you are expecting a null value appearing in left join and the property-accesses are able to accommodate that null value by being nullable type. _In such cases you are required to put nullable type cast. It is not a "work-around", it is required to materialize the results from the data you have._ Though if your query will never get null from left join (be careful as data can change over the type in an app), or you are not accessing a property on an entity which can be nullable due to left join then you shouldn't need it. The same exception can also arise when database contains null but client side property cannot take null values (even without left-joins) due to mismatch in model. There is also a possibility that there is a bug in EF Core somewhere. So if you have a query which has result types correct based on expected data and seeing this exception please file a new issue.

Thank you @smitpatel for this comment. It helped me in fixing my issue. 
Below is my left join linq query in which I was getting some null values from right side table(which was expected) which was causing the above error. After reading your comments I added nullable type in rightside table query and  null check in joining part of left and right table.

        return await (from items in _dbContext.LeftTable
                          join LeftJoinTable in ((from iss in _dbContext.RightTable
                                                    where iss.CId == cId && iss.ActionDate == null
                                                    select new { IId = **(int?)iss.IId** }).Distinct())
                          on **items.Id equals (LeftJoinTable.IId ?? 0)**  into ljoin
                          from LeftJoinTable in ljoin.DefaultIfEmpty()
                          orderby items.Name
                          select new CIDto
                          {
                              Id = items.Id,
                              Name = items.Name ,
                              IType = items.IType,
                              IsActive = (LeftJoinTable.IId ?? 0) == 0 ? 0 : 1 
                          }).AsNoTracking().ToListAsync();

**Comment by ajcvickers** [comment]
@Mijalski This issue is in the 6.0 milestone, which means we plan to fix it for EF Core 6.0.

**Comment by StrangeWill** [comment]
> Still happens with MetadataGenerator v1.2.0 and the latest EFC packages (5.0.1).
> 
> However from what I understand, this is now the expected behaviour and we'll have to add explicit "workarounds" to all left-joined queries like the one @shadowfoxish mentioned?
 
> @Mijalski This issue is in the 6.0 milestone, which means we plan to fix it for EF Core 6.0.

Yeah I'm a little confused here on what to expect in EF 6, are we going to expect the original behavior to come back in EF Core 6.0? We have a code base that we could add a _ton_ of conditional checks for 3-4 relationships deep but that would make complete mess of our selectors (if only we could do null conditionals in expression trees...), so we're effectively on .NET 5.0 with EF Core 3.1 (this isn't the first project we had to do this with).

An example of one of our selectors:
```csharp
  Procedure = new
  {
      cf.Field.Procedure.Name,
      cf.Field.Procedure.FollowUpInterval
  },
```

Checking cf, field and procedure each time is a bit rough.

Are we expecting EF Core 3.1 code to be compatible with EF 6?

**Comment by smitpatel** [comment]
@StrangeWill - This is what will happen in EF Core 6.0

- If the selector does conditional expression before construction of anonymous type, then we will not create instance of anonymous type
- If there is queryable of anonymous type which has DefaultIfEmpty applied over it then default value should be null rather than anonymous type with default property values (more complicated case of first one only)
- The selector you wrote above, if you have anonymous type construction without null check and if it ends up with null property value where expected is non-null value then it is going to throw the same error. I have explained in https://github.com/dotnet/efcore/issues/22517#issuecomment-758977488 things to consider when doing left join and why behavior is this way.
- None of above are going to be same behavior as EF Core 3.1 as EF Core 3.1 behavior was also wrong. It put wrong value rather than throwing exception when incorrect query results encountered.

Finally, this cannot be backported to 5.0 due to complexity of the fix.

**Comment by optiks** [comment]
We've just hit this after (attempting) to upgrade from 3.1 to 5.0.

Suggestion: Mention this is [Breaking changes in EF Core 5.0](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-5.0/breaking-changes)? 

Googling stack traces isn't much fun :)

**Comment by eric-miller-cdfw** [comment]
It's too bad LINQ expressions don't support the null conditional operator **?.** Then we could indicate a possible null in the navigation to the EF query and quiet the VS Analyzers that complain about a possible null dereference (CS8602).

```csharp
.Select(x => new {
    x.NavProperty1?.NavProperty2?.NonNullValueTypeProperty,
})
```

**Comment by Shiko1st** [comment]
This is definitely a breaking change

**Comment by smitpatel** [comment]
Related #19095

**Comment by channeladam** [comment]
This is very inconvenient, and has been 'punted' for 2 versions now...

**Comment by cyril265** [comment]
> So, as a user, you should carefully evaluate queries and see if you are expecting a null value appearing in left join and the property-accesses are able to accommodate that null value by being nullable type. In such cases you are required to put nullable type cast. It is not a "work-around", it is required to materialize the results from the data you have. Though if your query will never get null from left join (be careful as data can change over the type in an app), or you are not accessing a property on an entity which can be nullable due to left join then you shouldn't need it. The same exception can also arise when database contains null but client side property cannot take null values (even without left-joins) due to mismatch in model. There is also a possibility that there is a bug in EF Core somewhere. So if you have a query which has result types correct based on expected data and seeing this exception please file a new issue.

It doesn't work if the target type is nullable but contains non nullable properties. Also there is no point in being forced to handle C# null safety inside a SQL query which doesn't have such concept. This turns what would have been a pretty standard SQL query with a couple of left joins into this mess:


```C#
            query.Select(user => new
            {
                User = user,
                Person = user.Person!,
                user.Substitute,
                user.Substitute!.OtherUser,
                user.Substitute.ExternalUser
            })
            .Select(record => new UserRecord
            {
                Id = record.User.Id,
                Firstname = record.Person!.Firstname,
                Lastname = record.Person.Lastname,
                Email = record.Person.Email,
                Salutation = record.Person.Salutation,
                Substitute = record.Substitute != null
                    ? new SubstituteRecord
                    {
                        Type = record.Substitute.Type,
#pragma warning disable S3358
                        OtherUser = record.OtherUser != null
                            ? new()
                            {
                                Id = record.OtherUser.Id,
                                Email = record.OtherUser.Person!.Email,
                                Salutation = record.OtherUser.Person.Email,
                                Firstname = record.OtherUser.Person.Firstname,
                                Lastname = record.OtherUser.Person.Lastname,
                            }
                            : null,
                        ExternalUser = record.ExternalUser != null
                            ? record.Substitute.ExternalUser
                            : null
                    }
                    : null
            });
```

**Comment by GeoMarkou** [comment]
We just upgraded to .Net 6 and this is an embarrassing complete disaster. There's no easy way to determine where this bug occurs other than scouring all 300+ LINQ expressions to see if we're using a leftjoin, and selecting on the result. This isn't even listed as a breaking change on the update page. The error message was completely useless and resulted in around 2 hours of debugging because I couldn't figure out what was going wrong (why would I expect a simple SELECT to be the cause of the error?). Furthermore, this behaviour is completely inconsistent even within the LINQ query itself, and doesn't follow expected C# paradigms. Consider the following query:

```c#
var userData = await (
    from u in db.user
    
    // left join
    join s in db.optional_user_settings_row
    on u.user_id equals s.user_id
    into leftjoin_u from u in leftjoin_u.DefaultIfEmpty()

    // s.suspended Doesn't error ???
    where u.user_id = UserIDToSearchFor && !s.suspended

    select new UserDataResult
    {
        name = u.display_name,
        // s.colourblind_enabled does error ???
        colourblind = s.colourblind_enabled,
    }
).ToArrayAsync();
```

This query errors because `s` is null, so it can't cast the resulting `bool?` into the expected not-null `bool`. However, why doesn't `s.suspended` in the `where` clause cause an error? Logically, the actual `null` value is `s`, but this seems to be behaving against expectations, as though we've just got a completely blank `Object` instead of a row, and everything inside of it is evaluating to `null`, but only sometimes. Not only that, but it's not even evaluating to `default` as it should be based on the `DefaultIfEmpty()` name. This makes no sense at all, and honestly we're seriously considering completely abandoning using Entity Framework LINQ because this isn't the first time an update has broken queries across our application with no warning or logical way to fix it. I ended up coming to the same workaround conclusions of casting to `(bool?)` or using `== true` but again this is crazy behaviour to any user and causes confusion.

I know Microsoft doesn't take suggestions, and this is a pipe dream, but the following syntax is what I would actually expect LINQ to work like.

```c#
var userData = await (
    from u in db.user
    
    // left join - new syntax
    // `s` is highlighted as POSSIBLY NULL
    // and will display a warning in Visual Studio if accessed directly below
    left join s in db.optional_user_settings_row
    on u.user_id equals s.user_id

    // Use null propagating operator since `s` is the actual null value
    // and we can EXPLICITLY HANDLE IT instead of automatically doing weird stuff
    where u.user_id = UserIDToSearchFor && (s?.suspended ?? false)

    select new UserDataResult
    {
        display_name = u.display_name,
        // Use coalescing to decide what the default should be
        colourblind_enabled = s?.colourblind_enabled ?? false,
    }
).ToArrayAsync();
```

**Comment by blogcraft** [comment]
Same problem in EF 7

**Comment by darcast76** [comment]
I can't believe that after years, this error is still present.
I have tons of LINQ queries that work correctly in .NET Framework 3.1, and I'm not sure how they might behave after an update. 
Has anyone found any workarounds? Is there at least one way to obtain a compile-time error instead of a runtime error?

**Comment by dotVip** [comment]
I have some issue when I move app from EF 2 to EF 7.
I find solution that work for me

.SelectMany(d => d.Posts.DefaultIfEmpty(), (x, y) => new myModel
                              {
                                   ......
                                      somedata = y.somedata .GetType() == null ? "no data"  : y.somedata,
                                  ......
                             });

**Comment by junejasaket** [comment]
This issue is a significant blocker for our transition from EF Core 3.1 to later versions. It fundamentally changes the expected behavior of left joins involving nullable types, which would necessitate extensive modifications in our codebase to introduce null checks to avoid these exceptions. This isn't just a minor inconvenience but a substantial overhaul of existing, stable code.

A configuration option to toggle this behavior would be highly beneficial, offering a compromise. This would allow teams who rely on the pre-5.0 behavior to upgrade without rewriting significant portions of their code, while others who prefer the new behavior could opt-in. Without such an option, we're effectively stuck, unable to leverage the advancements in newer EF Core versions.

It's crucial to address this issue to facilitate smoother transitions and maintain backward compatibility, especially for complex, enterprise-level applications.

**Comment by Luigi6821** [comment]
Hi,
I believe that this is duplicate of same issue I already reported [#30915] (https://github.com/dotnet/efcore/issues/30915)
This is definetively a blocker issue for moving from EF to EF Core. I have many LINQ query that raise this issue.

Regards
Luigi

**Comment by ixetanet** [comment]
Really hoping for a fix on this one...

**Comment by Luigi6821** [comment]
> Really hoping for a fix on this one...

Please vote it

**Comment by mmanovic** [comment]
It is still present on EF 9.x. Any info when it will be resolved?

**Comment by luigi68** [comment]
I want post again this issue hoping that at least it will be solved on next releases

# EF Core Bug Report: DefaultIfEmpty Not Properly Handling Null Values in Left Joins

## Description
When using LINQ's `DefaultIfEmpty()` method with EF Core to create LEFT JOIN queries, the expected behavior is that when no matching records are found in the right table, the result should be NULL. However, the current implementation doesn't correctly handle this scenario, requiring manual null checks against the foreign key instead of relying on the expected behavior of DefaultIfEmpty.

## Expected Behavior
When using the standard LINQ pattern for a left join with `DefaultIfEmpty()`, null categories should be properly returned for addresses that don't have a matching category:

Example:
var addressesAnon = (from address in dbContext.Set<AMOS_ADDRESS>().AsNoTracking()
                     join c in dbContext.Categories
                     on address.ADDRESSCATEGORYID equals c.Id into categoryGroup
                     from category in categoryGroup.DefaultIfEmpty()
                     select new
                     {
                         Id = address.ADDRESSID,
                         Category = category // Should be null when no match exists
                     }).ToArray();

## Actual Behavior
The `Category` property does not correctly resolve to null when there is no matching category. The query does not properly translate the `DefaultIfEmpty()` semantics to SQL's LEFT JOIN behavior.

## Workaround
Currently, the only reliable solution is to explicitly check the foreign key for null is:

var addressesAnon = (from address in dbContext.Set<AMOS_ADDRESS>().AsNoTracking()
                     join c in dbContext.Categories
                     on address.ADDRESSCATEGORYID equals c.Id into categoryGroup
                     from category in categoryGroup.DefaultIfEmpty()
                     select new
                     {
                         Id = address.ADDRESSID,
                         Category = address.ADDRESSCATEGORYID == null ? null : category
                     }).ToArray();

## Reproduction Steps
1. Create two entities with a one-to-many relationship (e.g., Address and Category)
2. Add some addresses without corresponding categories (NULL foreign key)
3. Execute a LINQ query using the DefaultIfEmpty pattern as shown in the example
4. Observe that null handling doesn't work as expected

## Environment
- EF Core Version: EF Core 9.0.4
- Database Provider: SQL Server/In Memory
- .NET Version: NET 8
- Operating System: Windows

## Additional Information
This issue impacts a core LINQ operation that's commonly used for LEFT JOIN scenarios, forcing developers to use additional check for what should be standard functionality.

**Comment by adrius** [comment]
> I want post again this issue hoping that at least it will be solved on next releases
> 
> # EF Core Bug Report: DefaultIfEmpty Not Properly Handling Null Values in Left Joins
> ## Description
> When using LINQ's `DefaultIfEmpty()` method with EF Core to create LEFT JOIN queries, the expected behavior is that when no matching records are found in the right table, the result should be NULL. However, the current implementation doesn't correctly handle this scenario, requiring manual null checks against the foreign key instead of relying on the expected behavior of DefaultIfEmpty.
> 
> ## Expected Behavior
> When using the standard LINQ pattern for a left join with `DefaultIfEmpty()`, null category should be properly returned for addresses that don't have a matching category:
> 
> Example: var addressesAnon = (from address in dbContext.Set<AMOS_ADDRESS>().AsNoTracking() join c in dbContext.Categories on address.ADDRESSCATEGORYID equals c.Id into categoryGroup from category in categoryGroup.DefaultIfEmpty() select new { Id = address.ADDRESSID, Category = category // Should be null when no match exists }).ToArray();
> 
> ## Actual Behavior
> The `Category` property does not correctly resolve to null when there is no matching category. The query does not properly translate the `DefaultIfEmpty()` semantics to SQL's LEFT JOIN behavior.
> 
> ## Workaround
> Currently, the only reliable solution is to explicitly check the foreign key for null is:
> 
> var addressesAnon = (from address in dbContext.Set<AMOS_ADDRESS>().AsNoTracking() join c in dbContext.Categories on address.ADDRESSCATEGORYID equals c.Id into categoryGroup from category in categoryGroup.DefaultIfEmpty() select new { Id = address.ADDRESSID, Category = address.ADDRESSCATEGORYID == null ? null : category }).ToArray();
> 
> ## Reproduction Steps
> 1. Create two entities with a one-to-many relationship (e.g., Address and Category)
> 2. Add some addresses without corresponding categories (NULL foreign key)
> 3. Execute a LINQ query using the DefaultIfEmpty pattern as shown in the example
> 4. Observe that null handling doesn't work as expected
> 
> ## Environment
> * EF Core Version: EF Core 9.0.4
> * Database Provider: SQL Server/In Memory
> * .NET Version: NET 8
> * Operating System: Windows
> 
> ## Additional Information
> This issue impacts a core LINQ operation that's commonly used for LEFT JOIN scenarios, forcing developers to use additional check for what should be standard functionality. Please note that even if I use the workaround I have issue when I do the following:
> 
> var addressesAnon = (from address in dbContext.Set<AMOS_ADDRESS>().AsNoTracking() join c in dbContext.Categories on address.ADDRESSCATEGORYID equals c.Id into categoryGroup from category in categoryGroup.DefaultIfEmpty() select new { Id = address.ADDRESSID, Category = address.ADDRESSCATEGORYID == null ? null : category });
> 
> var addressWithCatNull = addressAnon.Where(a => a.Category == null)
> 
> _Error:_ System.InvalidOperationException HResult=0x80131509 Message=The LINQ expression 'DbSet<AMOS_ADDRESS>() .LeftJoin( inner: DbSet<AMOS_ADDRESSCATEGORY>(), outerKeySelector: a => (object)a.ADDRESSCATEGORYID, innerKeySelector: a0 => (object)(decimal?)a0.ADDRESSCATEGORYID, resultSelector: (a, a0) => new TransparentIdentifier<AMOS_ADDRESS, AMOS_ADDRESSCATEGORY>( Outer = a, Inner = a0 )) .Where(ti => !(ti.Outer.ADDRESSCATEGORYID != null) || new Category{ Id = ti.Inner.ADDRESSCATEGORYID, Name = ti.Inner.DESCRIPTION } == null)' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information. Source=Microsoft.EntityFrameworkCore StackTrace: at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.Translate(Expression expression) at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutorExpression[TResult](Expression query) at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query) at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async) at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async) at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass11_0`1.<ExecuteCore>b__0() at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler) at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteCore[TResult](Expression query, Boolean async, CancellationToken cancellationToken) at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query) at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression) at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetEnumerator() at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection) at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source) at ReproducibleNullObject.Tests.RunInMemory() in C:\Users\luigi.esposito\Desktop\Repos\ReproducibleNullObject\ReproducibleNullObject\RunInMemory.cs:line 145 at Program.
> 
> $(String[] args) in C:\Users\luigi.esposito\Desktop\Repos\ReproducibleNullObject\ReproducibleNullObject\Program.cs:line 3
> For workarounding it I need to rewrite query as follows: var addressWithCatNull = addressAnon.Where(a => a.Category.Id == null) For which I have warning CS0472 (https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0472?f1url=%3FappId%3Droslyn%26k%3Dk(CS0472))

Great sum-up, we're facing something similar and would be great if this could be addressed

**Comment by shadowfoxish** [comment]
@luigi68 @adrius 
This... seems like a separate issue, unrelated to the nullable type materialization problem this issue is about.

> ... could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'.

Have you tried something like this: _(I think you have to check the actual type that's being materialized for null, not its foreign-key-reference counterpart. I also added AsNoTracking on the Categories entity. I'm also confused why you're using .Set<AMOS_ADDRESS> instead of whatever object that maps to on the context. Visually there seems to be some weirdness with this code.)_

```c#
var addressesAnon = (from address in dbContext.Set<AMOS_ADDRESS>().AsNoTracking()
join c in dbContext.Categories.AsNoTracking() on address.ADDRESSCATEGORYID equals c.Id into categoryGroup
from category in categoryGroup.DefaultIfEmpty()
select new
{
    Id = address.ADDRESSID,
    Category = category == null ? null : category
});
```

If this were my code base, I would desire to have the foreign keys specified in the DbContext with proper navigation properties using .HasMany .WithOne, etc, and then the query becomes a bit cleaner as well.

```c#
var addressesAnon = (from address in dbContext.AmosAddresses.AsNoTracking()
from category in address.Categories //Assuming this is a 1:many relationship
select new
{
    Id = address.ADDRESSID,
    Category = category == null ? null : category
});
```

**Comment by cyril265** [comment]
Can we at least get a better error message if this bug is not getting fixed?
It should contain the property name. As it is right now there is no way to tell which part of the select is the issue without commenting out stuff until it works.

**Comment by DhamodharanSiva** [comment]
Hello @roji / @AndriySvyryd,

Our project is also impacted because of this issue. Will it be fixed in 11.0 release scheduled to be released in Nov 2026?

**Comment by roji** [comment]
It's on the plan (see the milestone on this issue) - I really hope we get to it.

**Comment by roji** [comment]
# AI Triage

The below is an AI-generated analysis and may contain inaccuracies.

**Issue type:** Bug (confirmed)
**Area labels:** `area-query` (already applied)
**Regression:** Yes — this was introduced in EF Core 5.0 preview 8 as a side effect of the fix for https://github.com/dotnet/efcore/issues/20633. Prior to that fix, EF Core returned incorrect results (an instance with default property values instead of `null`); after the fix, it throws instead.

## Bug Confirmed

Reproduced on **EF Core 10.0.0-preview.3** with **SQLite**, using the new `LeftJoin` LINQ operator. The generated SQL is correct — a proper `LEFT JOIN` is produced:

```sql
SELECT "b"."Url", "p"."PostId", "p"."BlogId", "p"."Content"
FROM "Blogs" AS "b"
LEFT JOIN "Posts" AS "p" ON "b"."BlogId" = "p"."BlogId"
```

However, when the right side of the join returns `NULL` columns (i.e. the blog has no matching posts), materialization fails with:

```
System.InvalidOperationException: Nullable object must have a value.
   at lambda_method53(Closure, QueryContext, DbDataReader, ResultContext, SingleQueryResultCoordinator)
```

The core problem is that EF Core's query pipeline does not detect that the entire right-side entity/projection should be `null` when all its columns are `NULL` from a left join. Instead, it tries to materialize the entity and fails when assigning `NULL` to non-nullable value-type properties (e.g. `int PostId`).

<details>
<summary>Minimal repro</summary>

Create a new console project with `Microsoft.EntityFrameworkCore.Sqlite` 10.0.0-preview.3 (or any 5.x+ version) and the following `Program.cs`:

```csharp
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

await using var context = new TestContext();
await context.Database.EnsureDeletedAsync();
await context.Database.EnsureCreatedAsync();

// Seed a blog with no posts
context.Blogs.Add(new Blog { Url = "http://example.com" });
await context.SaveChangesAsync();

var dbPosts = from p in context.Posts
              select new PostDto
              {
                  PostId = p.PostId,
                  BlogId = p.BlogId,
                  Content = p.Content
              };

var query = context.Blogs
    .LeftJoin(
        dbPosts,
        blog => blog.BlogId,
        post => post.BlogId,
        (blog, post) => new
        {
            blog.Url,
            Post = post
        });

// Throws: System.InvalidOperationException: Nullable object must have a value.
var results = await query.ToListAsync();

foreach (var r in results)
{
    Console.WriteLine($"Blog: {r.Url}, Post: {(r.Post == null ? "null" : r.Post.PostId.ToString())}");
}

Console.WriteLine("Expected: Post should be null for blog with no posts");

public class TestContext : DbContext
{
    public DbSet<Blog> Blogs { get; set; }
    public DbSet<Post> Posts { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            .UseSqlite("Data Source=test.db")
            .LogTo(Console.WriteLine, LogLevel.Information)
            .EnableSensitiveDataLogging();
}

public class Blog
{
    public int BlogId { get; set; }
    public string Url { get; set; } = null!;
    public List<Post> Posts { get; set; } = [];
}

public class Post
{
    public int PostId { get; set; }
    public string Title { get; set; } = null!;
    public string Content { get; set; } = null!;
    public int BlogId { get; set; }
    public Blog Blog { get; set; } = null!;
}

public class PostDto
{
    public int PostId { get; set; }
    public int BlogId { get; set; }
    public string Content { get; set; } = null!;
}
```

</details>

## Possible Duplicates

- https://github.com/dotnet/efcore/issues/30915 — **Very likely duplicate.** Same `DefaultIfEmpty()` + left join materialization failure with projected sub-query. Currently open in Backlog.
- https://github.com/dotnet/efcore/issues/28273 — Closed as duplicate of this issue.
- https://github.com/dotnet/efcore/issues/26873 — Closed as duplicate of this issue. Also covers the related scenario where EF Core 3.1 materializes a non-null instance with default values instead of `null`.
- https://github.com/dotnet/efcore/issues/19095 — Noted as related by @smitpatel.

**Comment by roji** [comment]
Took another look at this, and a basic blocker to fixing this is our navigation expansion reordering of the operators. That is, the incoming query tree looks like this:

```c#
DbSet<RootEntity>()
    .LeftJoin(
        inner: DbSet<AssociateType>()
            .Select(a => new AssociateProjection{ 
                Id = a.Id, 
                Int = a.Int, 
                Name = a.Name 
            }
            ), 
        outerKeySelector: r => r.OptionalAssociateId, 
        innerKeySelector: p => (int?)p.Id, 
        resultSelector: (root, projection) => new { 
            Id = root.Id, 
            Name = root.Name, 
            Associate = projection
         })
```

The Select is nested within the LeftJoin argument; so the LeftJoin could see the projection to AssociateProjection, and apply a conditional with a null-check to ensure that the inner is null, rather than attempting to instantiate it though its scalars are NULL (the source of the exception).

However, our nav rewriting in preprocessing transforms to the following:

```c#
DbSet<RootEntity>()
    .LeftJoin(
        inner: DbSet<AssociateType>(), 
        outerKeySelector: r => (object)r.OptionalAssociateId, 
        innerKeySelector: a => (object)(int?)a.Id, 
        resultSelector: (r, a) => new TransparentIdentifier<RootEntity, AssociateType>(
            Outer = r, 
            Inner = a
        ))
    .Select(ti => new { 
        Id = ti.Outer.Id, 
        Name = ti.Outer.Name, 
        Associate = new AssociateProjection{ 
            Id = ti.Inner.Id, 
            Int = ti.Inner.Int, 
            Name = ti.Inner.Name 
        }
     })
```

The Select has been lifted out, so the LeftJoin can no longer see the Select. When translating this, we have no way of knowing that we should add a null check around `new AssociateProject()`, or what we'd property check in that null check (the inner's Id, based on the inner key selector in the LeftJoin). 

The good news is that there's work underway to stop rewriting the query tree in this way, tracked by #20291 - there's a good chance this will happen for 11.0; if it does, I'll try to tackle this problem as well.

What's important to understand, is that we absolutely should **not** materialize an AssociateProjection here but with a default Id (zero); that would be incorrect data (and was the pre-5.0 behavior, apparently).

**Comment by roji** [comment]
Also see #38055 when working on this.

**Comment by ajcvickers** [comment]
Follow up work after #38479:

All are sub-cases of the umbrella #22517 unless noted.

- [x] Second join applied after the left join, then whole-object projection (shaper rebuilt → marker lost). Repro: closed dup #26873.
- [x] `GroupBy` applied after the left join, then whole-object projection (shaper rebuilt via `ApplyGrouping`). Repro: closed dup #28119.
- [ ] ~~Plain inner with no pushdown (no aggregate/filter) via the `LeftJoin` operator — marker is inlined into the combined `SELECT` and never becomes `NULL`. *(no dedicated issue)*~~ → Tracked in #38608; **not fixed as part of this work.** The stated mechanism is a misdiagnosis: nav-expansion folds the plain inner `Select` into the pending selector, so the marker is never injected (not "inlined"). It can't be fixed in the relational layer — deferred to the nav-expansion rework (#32957 / #37859).
- [x] Value-type `struct` / `record struct` whole-object projection (members not defaulted; materialization throws).
- [x] Constructor-bound DTO / positional `record struct` / `ValueTuple` whole-object projection (fails at translation — a distinct path).
- [ ] Overlaps #28248.
- [ ] `Union` / set-operation of two left-join non-entity projections (set-operation translation failure). *(no dedicated issue)*
- [ ] Server-side `obj == null` / `obj != null` on a left-joined non-entity, in `Where` / `OrderBy` (translation failure). A complete #22517 (true null-tracking) would likely enable this. *(no dedicated issue)*
- [ ] `RightJoin` where the **outer** (non-entity) side is the nullable one — the marker is injected for the inner side only, so this still throws.
- [x] `Nullable<T>` whole-object projection (latent and currently unreachable; the gate skips on `IsValueType`, but `default(T?)` is `null`).
